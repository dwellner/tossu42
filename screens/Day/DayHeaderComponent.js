import React from "react";
import PropTypes from "prop-types";
import Formatters from "../../utils/Formatters";
import DateUtils from "../../utils/DateUtils";
import Texts from "../../constants/Texts";
import { StyleSheet, Text, View, Button } from "react-native";
import Colors from "../../constants/Colors";
import Styles from "../../constants/Styles";

NoDayProgram = ({ weekProgram }) => {
  const firstDate = Formatters.dateToDateLabel(
    weekProgram.weeks[0].days[0].date
  );
  const lastDate = Formatters.dateToDateLabel(
    weekProgram.weeks.slice(-1)[0].days.slice(-1)[0].date
  );

  return (
    <View style={styles.goalContainer}>
      <Text style={{ ...Styles.defaultContent, textAlign: "center" }}>
        {Texts.labels.programSetFor} {firstDate} - {lastDate}
      </Text>
    </View>
  );
};

const TargetEvent = ({ eventName }) => (
  <View style={styles.goalContainer}>
    <Text style={styles.text_targetName}>{eventName}</Text>
  </View>
);

const DaysUntilTargetEvent = ({ targetEvent, daysUntil }) => (
  <View style={styles.goalContainer}>
    <View style={{ flex: 1, flexDirection: "row" }}>
      <Text style={styles.text_days_until}>{daysUntil}</Text>
      <Text style={styles.text_label}>
        {daysUntil > 1 ? Texts.labels.daysUntil : Texts.labels.dayUntil}
      </Text>
    </View>
    <Text style={styles.text_targetName}>
      {targetEvent.name} - {Formatters.dateToDateLabel(targetEvent.date)}
    </Text>
  </View>
);

export default class DayHeaderComponent extends React.Component {
  static propTypes = {
    weekProgram: PropTypes.object.isRequired,
    date: PropTypes.string.isRequired,
    targetEvent: PropTypes.object.isRequired,
    changeDate: PropTypes.func.isRequired
  };

  render() {
    const { date, weekProgram, changeDate, targetEvent } = this.props;
    const daysUntil = DateUtils.difference(targetEvent.date, date) - 1;

    const prevWeek = () => changeDate(DateUtils.nextDate(date, -7));
    const nextWeek = () => changeDate(DateUtils.nextDate(date, 7));

    const targetComponent =
      daysUntil < 0 ? (
        <NoDayProgram weekProgram={weekProgram} />
      ) : daysUntil > 0 ? (
        <DaysUntilTargetEvent targetEvent={targetEvent} daysUntil={daysUntil} />
      ) : (
        <TargetEvent eventName={targetEvent.name} />
      );

    return (
      <View style={styles.component}>
        <View style={{ flex: 1, flexDirection: "row" }}>
          <View style={{ flex: 1 }}>
            <Button color="#333" title="◀" onPress={prevWeek} />
          </View>
          <View style={{ flex: 4, alignItems: "center" }}>
            <Text style={styles.text_date}>
              {Formatters.dateToDayLabel(date)}
            </Text>
          </View>
          <View style={{ flex: 1 }}>
            <Button color="#333" title="▶" onPress={nextWeek} />
          </View>
        </View>
        {targetComponent}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  component: {
    ...Styles.widgetContainer,
    paddingTop: 30,
    flex: 1,
    alignItems: "center",
    alignItems: "stretch"
  },

  goalContainer: {
    flex: 1,
    alignItems: "center"
  },

  text_days_until: {
    ...Styles.defaultContent,
    color: Colors.tintColor,
    fontWeight: "bold"
  },
  text_date: { ...Styles.defaultContent, fontSize: 24 },

  text_label: {
    ...Styles.defaultContent,
    marginLeft: 4
  },
  text_targetName: { ...Styles.defaultContent }
});
