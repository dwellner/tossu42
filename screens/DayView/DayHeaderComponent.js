import React from "react";
import PropTypes from "prop-types";
import Formatters from "../../utils/Formatters";
import DateUtils from "../../utils/DateUtils";
import Texts from "../../constants/Texts";
import { StyleSheet, Text, View, Button } from "react-native";

export default class DayHeaderComponent extends React.Component {
  static propTypes = {
    date: PropTypes.string.isRequired,
    targetEvent: PropTypes.object.isRequired,
    changeDate: PropTypes.func.isRequired
  };

  getTargetComponent(targetEvent, date) {
    const daysUntil = DateUtils.difference(targetEvent.date, date) - 1;

    if (daysUntil < 0) return undefined;

    if (daysUntil > 0)
      return (
        <View
          style={{
            flex: 1,
            padding: 10,
            borderWidth: 1,
            borderTopColor: "#333",
            alignItems: "center",
            backgroundColor: "#1117"
          }}
        >
          <View style={{ flex: 1, flexDirection: "row" }}>
            <Text style={styles.text_days_until}>{daysUntil}</Text>
            <Text style={styles.text_label}>
              {daysUntil > 1 ? Texts.labels.daysUntil : Texts.labels.dayUntil}
            </Text>
          </View>
          <Text style={styles.text_targetName}>
            {targetEvent.name} - {Formatters.dateToDateLabel(targetEvent.date)}
          </Text>
        </View>
      );

    return (
      <View
        style={{
          flex: 1,
          padding: 10,
          borderWidth: 1,
          borderTopColor: "#333",
          alignItems: "center",
          backgroundColor: "#1117"
        }}
      >
        <Text style={styles.text_targetName}>
          {targetEvent.name}
        </Text>
      </View>
    );
  }

  render() {
    const { date, targetEvent, changeDate } = this.props;

    const targetComponent = this.getTargetComponent(targetEvent,date);
    return (
      <View style={styles.component}>
        <View
          style={{ flex: 1, flexDirection: "row", backgroundColor: "#1115" }}
        >
          <View style={{ flex: 1 }}>
            <Button
              color="#333"
              title="◀"
              onPress={() => changeDate(DateUtils.nextDate(date, -7))}
            />
          </View>
          <View style={{ flex: 4, alignItems: "center" }}>
            <Text style={styles.text_date}>
              {Formatters.dateToDayLabel(date)}
            </Text>
          </View>
          <View style={{ flex: 1 }}>
            <Button
              color="#333"
              title="▶"
              onPress={() => changeDate(DateUtils.nextDate(date, 7))}
            />
          </View>
        </View>
        {targetComponent}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  component: {
    flex: 1,
    alignItems: "center",
    alignItems: "stretch"
  },
  text_days_until: {
    color: "#f442df",
    fontSize: 24,
    fontWeight: "bold"
  },
  text_date: {
    color: "#fff",
    fontSize: 24
  },
  text_label: {
    color: "#fff",
    textAlignVertical: "center",
    fontSize: 24,
    marginLeft: 4
  },
  text_targetName: {
    color: "#fff",
    fontSize: 24,
    fontWeight: "900"
  }
});
