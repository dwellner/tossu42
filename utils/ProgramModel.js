import DateUtils from "./DateUtils";
import ProgramService from "../data/ProgramService";

const getIntensityFactor = runType =>
  ({
    iv: 2.5,
    m: 2,
    r: 2,
    t: 2,
    k: 1.5,
    pi: 1.5,
    pe: 1.5,
    ve: 0.5
  }[runType] || 1);

const getIVTotalDistance = (ivDistance, repeats) => {
  const warmup = 2,
    coolDown = 1,
    recovery = 0.4;
  return warmup + (ivDistance + recovery) * repeats + coolDown;
};

const getDayDistance = day => {
  if (day.type === "iv") return getIVTotalDistance(day.distance, day.repeat);
  if (day.type === "lepo") return 0;
  return day.distance;
};

const getDayIntensity = day =>
  getIntensityFactor(day.type) * getDayDistance(day);

const getWeekDistance = week =>
  week.map(getDayDistance).reduce((a, b) => a + b);

const getIntensity = week => week.map(getDayIntensity).reduce((a, b) => a + b);

const getStretchedProgramStart = (program, programLength, totalLength) => {
  let repeats = [];

  const stretchLengthInWeeks = Math.ceil(totalLength - programLength);
  const repeatBlock = program.weeks.slice(0, program.stretchRules.repeat);
  const numberOfRepeatBlocks = Math.ceil(
    stretchLengthInWeeks / repeatBlock.length
  );
  for (i = 0; i < numberOfRepeatBlocks; i++) {
    repeats = repeats.concat(repeatBlock.map(days => [].concat(days)));
  }
  const excessWeeks = repeats.length - stretchLengthInWeeks;
  return repeats.slice(excessWeeks);
};

const getProgramWeeks = (program, length) => {
  const programLength = program.weeks.length;

  if (length === programLength || program.stretchRules === undefined) {
    return program.weeks;
  }

  if (length > programLength) {
    const stretchedStart = getStretchedProgramStart(
      program,
      programLength,
      length
    );
    return [].concat(stretchedStart, program.weeks);
  }

  // TODO: check total length is within program range, or return mimimum
  return []; // TODO: handle shrinking
};

export default {
  getWeekProgram: (targetDate, { targetTime, name, length }) => {
    if (targetDate === null) return null;
    
    const programs = ProgramService.getPrograms(targetTime, name, length);
    if (programs.length === 0) return null;
    const program = programs[0];

    const programWeeks = getProgramWeeks(program, length);
    const firstDate = DateUtils.nextDate(
      targetDate,
      programWeeks.length * -7 + 1
    );
    const maxWeekDistance = programWeeks
      .map(getWeekDistance)
      .reduce((a, b) => Math.max(a, b), 0);

    const maxIntensity = programWeeks
      .map(getIntensity)
      .reduce((max, w) => Math.max(max, w), 0);

    let i = 0;
    const weeks = programWeeks.map(week => {
      const distance = getWeekDistance(week);
      const intensityLevel = getIntensity(week) / maxIntensity;
      const distanceLevel = distance / maxWeekDistance;
      const days = week.map(day => ({
        ...day,
        date: DateUtils.nextDate(firstDate, i++)
      }));
      return { distance, intensityLevel, distanceLevel, days };
    });
    return { weeks };
  }
};
