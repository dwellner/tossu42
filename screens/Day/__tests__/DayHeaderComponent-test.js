import "react-native";
import React from "react";
import DayHeaderComponent from "../DayHeaderComponent";
import renderer from "react-test-renderer";
import ProgramModel from "../../../utils/ProgramModel";

it("renders correctly", () => {
  const targetEvent = { name: "Foo maraton", date: "2018-10-30" };

  const weekProgram = ProgramModel.getWeekProgram("2018-10-30", {
    targetTime: 240,
    name: "Tossu.com 4 tunnin t√§ysohjelma",
    length: 24
  });

  const week = weekProgram.weeks[0];
  const tree = renderer
    .create(
      <DayHeaderComponent
        weekProgram={weekProgram}
        date={week.days[0].date}
        targetEvent={targetEvent}
        changeDate={() => {}}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
