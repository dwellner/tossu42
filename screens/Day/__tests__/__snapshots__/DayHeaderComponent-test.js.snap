// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "#1119",
      "borderBottomWidth": 1,
      "borderColor": "#333",
      "borderTopWidth": 1,
      "flex": 1,
      "marginBottom": 16,
      "paddingBottom": 16,
      "paddingTop": 30,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#333",
                },
              ]
            }
          >
            ◀
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 4,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
          }
        }
      >
        Keskiviikko 16.5.
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#333",
                },
              ]
            }
          >
            ▶
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#f442df",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        167
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 20,
            "marginLeft": 4,
          }
        }
      >
        päivää ennen
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#fff",
          "fontSize": 20,
        }
      }
    >
      Foo maraton
       - 
      30.10.18
    </Text>
  </View>
</View>
`;
